### PR TITLE
Add DataPayload::dynamic_cast with example

### PR DESCRIPTION
**Pull request by sffc**
I came across this in the context of the transliteration example. Adding this function makes forking providers easier to write.

### CODE REVIEW

**Review by Manishearth — APPROVED** [review]

File: provider/core/src/response.rs [hunk 1]
```diff
@@ -718,6 +718,81 @@ where
...
+    ///
+    /// impl<M, P0, P1> DataProvider<M> for MyForkingProvider<P0, P1>
+    /// where
+    ///     M: DataMarker,
+    ///     P0: DataProvider<M>,
+    ///     P1: DataProvider<HelloWorldV1Marker>,
+    /// {
+    ///     #[inline]
+    ///     fn load(&self, req: DataRequest) -> Result<DataResponse<M>, DataError> {
+    ///         if TypeId::of::<HelloWorldV1Marker>() == TypeId::of::<M>() {
```
Comment (robertbastian) [inline provider/core/src/response.rs:755]: Everywhere we compare data markers types we compare `M::INFO`. Comparing `TypeId::of::<M>()` is a new way to do this and we should agree on one.

**Review by robertbastian — COMMENTED** [review]
This should be applied to the tutorials: https://github.com/unicode-org/icu4x/blob/1426a7e5a5e63cb4dde1f0b97e4a6000a090c9e7/tutorials/data_provider.md?plain=1#L218-L219

**Comment by sffc** [comment]
> This should be applied to the tutorials:

`downcast_mut` and `dynamic_cast_mut` still exist. This PR adds `dymamic_cast`, which is behavior that didn't exist in the library before.

File: provider/core/src/response.rs (hunk 1 above)
Comment (sffc) [reply provider/core/src/response.rs:755]: Hmm, interesting observation.

**Comment by robertbastian** [comment]
Right, the tutorial should use `dynamic_cast_mut`

**Comment by robertbastian** [comment]
https://github.com/unicode-org/icu4x/pull/5490

**Pull request merged**